### ISSUE TITLE
Programmatic multiple interceptor registration is possible?

### ISSUE THREAD

**Issue by brkmustu** [z4kn4fein/stashbox#68]
Hi Peter,

i have a programmatic method for interceptor registration;

        protected static void RegisterWithInterceptor<TService,TImplementation,TIntercetor>(StashboxContainer container)
            where TIntercetor : class, IInterceptor
            where TImplementation : class, TService
            where TService : class
        {
            var proxyBuilder = new DefaultProxyBuilder();
            var proxy = proxyBuilder
                .CreateInterfaceProxyTypeWithTargetInterface(
                     typeof(TService),
                     new Type[0],
                     ProxyGenerationOptions.Default);

            container.Register<IInterceptor, TIntercetor>()
                 .Register<TService, TImplementation>()
                 .RegisterDecorator<TService>(proxy);
        }

currently I can record multiple interceptors without having a problem when following a path like

            var proxyBuilder = new DefaultProxyBuilder();
            var proxy = proxyBuilder
                .CreateInterfaceProxyTypeWithTargetInterface(
                     typeof(IReportAdapter),
                     new Type[0],
                     ProxyGenerationOptions.Default);

            container.Register<IInterceptor, AuthenticationInterceptor>()
                 .Register<IInterceptor, LoggingInterceptor>()
                 .Register<IReportAdapter, ReportAdapter>()
                 .RegisterDecorator<IReportAdapter>(proxy);

is it possible to approach such a method as follows?

        protected static void RegisterWithInterceptor<TService, TImplementation>(StashboxContainer container, params IInterceptor[] interceptors)
            where TImplementation : class, TService
            where TService : class
        {
            var proxyBuilder = new DefaultProxyBuilder();
            var proxy = proxyBuilder
                .CreateInterfaceProxyTypeWithTargetInterface(
                     typeof(TService),
                     new Type[0],
                     ProxyGenerationOptions.Default);

            container.Register<IInterceptor>(interceptors)
                 .Register<TService, TImplementation>()
                 .RegisterDecorator<TService>(proxy);
        }

or is there any other way to register more than one intercetor for an interface?

**Comment by z4kn4fein** [comment]
Hey,

Sure, I think you are searching for the `RegisterTypesAs()` registration method. You can use it like this:
```
protected static void RegisterWithInterceptor<TService, TImplementation>(StashboxContainer container, params IInterceptor[] interceptors)
        where TImplementation : class, TService
        where TService : class
 {
      var proxyBuilder = new DefaultProxyBuilder();
      var proxy = proxyBuilder
          .CreateInterfaceProxyTypeWithTargetInterface(
               typeof(TService),
               new Type[0],
               ProxyGenerationOptions.Default);

      container.RegisterTypesAs<IInterceptor>(interceptors)
           .Register<TService, TImplementation>()
           .RegisterDecorator<TService>(proxy);
}
```

**Comment by brkmustu** [comment]
When you copy and paste the code you share it gives syntax error. do I need to install an additional package for this?

**Comment by z4kn4fein** [comment]
Oh sorry, I misunderstood your example, those are actual interceptor implementations in the params arguments. Currently, multiple instance registration is not supported but I can add it in the next version.

In the meantime and if it's acceptable, you can still use the `RegisterTypesAs()` api if you are passing the types instead of the actual instances like this:
```
protected static void RegisterWithInterceptor<TService, TImplementation>(StashboxContainer container, params Type[] interceptorTypes)
        where TImplementation : class, TService
        where TService : class
 {
      var proxyBuilder = new DefaultProxyBuilder();
      var proxy = proxyBuilder
          .CreateInterfaceProxyTypeWithTargetInterface(
               typeof(TService),
               new Type[0],
               ProxyGenerationOptions.Default);

      container.RegisterTypesAs<IInterceptor>(interceptorTypes)
           .Register<TService, TImplementation>()
           .RegisterDecorator<TService>(proxy);
}
```

Or if you'd like to stay with the instance version, then you can iterate on them and call the `RegisterInstanceAs()` function of the container like this:
```
protected static void RegisterWithInterceptor<TService, TImplementation>(StashboxContainer container, params IInterceptor[] interceptors)
        where TImplementation : class, TService
        where TService : class
 {
      var proxyBuilder = new DefaultProxyBuilder();
      var proxy = proxyBuilder
            .CreateInterfaceProxyTypeWithTargetInterface(
                 typeof(TService),
                 new Type[0],
                 ProxyGenerationOptions.Default);

       foreach (var interceptor in interceptors)
            container.RegisterInstanceAs(interceptor);

       container.Register<TService, TImplementation>()
            .RegisterDecorator<TService>(proxy);
}
```

**Comment by z4kn4fein** [comment]
In the meantime I also added a `RegisterInstancesAs()` functionality so you can use that like this:
```
container.RegisterInstancesAs<IInterceptor>(interceptors)
           .Register<TService, TImplementation>()
           .RegisterDecorator<TService>(proxy);
```

**Comment by mhmtmustu** [comment]
Thank you for your interest, Peter.
I wish you success in your work...

In addition, I have another fixation :)
i will share it at the first opportunity

**Comment by brkmustu** [comment]
Thank you for your interest, Peter.
I wish you success in your work...

In addition, I have another fixation :)
i will share it at the first opportunity

**Comment by z4kn4fein** [comment]
released in [v2.8.6](https://github.com/z4kn4fein/stashbox/releases/tag/2.8.6)

I'm closing this now, please create a new ticket for your other issue, thanks! :)
